doc/rules/phpdoc/phpdoc_list_type.rst

```diff
@@ -0,0 +1,65 @@
+=========================
+Rule ``phpdoc_list_type``
+=========================
+
+PHPDoc list types must be written in configured style.
+
+Configuration
+-------------
+
+``style``
+~~~~~~~~~
+
+Whether to use ``array`` or ``list`` as type.
+
+Allowed values: ``'array'`` and ``'list'``
+
+Default value: ``'list'``
+
+Examples
+--------
+
+Example #1
+~~~~~~~~~~
+
+*Default* configuration.
+
+.. code-block:: diff
+
+   --- Original
+   +++ New
+    <?php
+    /**
+   - * @param bool[] $a
+   - * @param array<int> $b
+   + * @param list<bool> $a
+   + * @param list<int> $b
+     * @param array<string, int> $c
+     * @param list<int> $d
+     */
+
+Example #2
+~~~~~~~~~~
+
+With configuration: ``['style' => 'array']``.
+
+.. code-block:: diff
+
+   --- Original
+   +++ New
+    <?php
+    /**
+   - * @param bool[] $a
+   + * @param array<bool> $a
+     * @param array<int> $b
+     * @param array<string, int> $c
+   - * @param list<int> $d
+   + * @param array<int> $d
```

Comment:
this is not a list. it's `array<array-key, int>`

Reply:
👎🏼 for removing `['style' => 'array']`, see https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/7077

maybe that option should not touch `list<>` and only change `[]` syntax 🤔

Reply:
Yes, the `array|list` should be used only when converting from `[]` syntax.

Reply:
It also should be possible to convert from `array` to `list`, should that be a separate option or included in `list` one?

Reply:
> suggest removing this `['style' => 'array']` option

👍 plus the fixer should be renamed to reflect only the https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/7077 (ie. `xxx[]` to `array<xxx>`)

converting `xxx[]` to `list<xxx>` is really not possible by a fixer as `[]` does simply not imply a list

Reply:
> is really not possible by a fixer (without static analysis knowledge)

As in many other fixers. If someone has `[]` and wants to convert, then in many cases it is in fact a `list`, not an `array`.

In this very repository if you apply this fixer with default config, it will change 112 files with all changes to `list` are valid. The ones that were actually arrays were fixed [here](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/pull/7779) - that's 17 files.

Reply:
I strongly disagree. `xxx[]` has a meaning. The meaning is an array with any keys. The fixer must hold this knowledge and even 1% of usecases will be enough, in your statement, you are saying the fixer is wrong in more than 10% of usecases!

Reply:
What did you disagree exactly about? That 112 is greater than 17? 😉 I showed practical usage of `list` option, if you do not like it, that's totally fine, that is why there is an option `array`, exactly for you.

Please, think about the other people. The fact that you don't want to change `[]` to `list<>` does not mean the whole world thinks the same. Converting `[]` to `array<type, value>` or `list<value>` - as [mentioned by a "random" person](https://github.com/PHP-CS-Fixer/PHP-CS-Fixer/issues/4698#issuecomment-1874661083) would be much easier with having `list` option.

Reply:
Let me explain it.

`xxx[]` is 1:1 equivalent to `array<xxx>`. So there should be a fixer for this with this behaviour as a default config. Even for 1% of usecases we cannot/should not change, even phpdoc, to something wrong by default.

Then we have `array<xxx>`. There can be a \*different\* and \*risky\* fixer to fix such syntax to `list<xxx>`.

Does this reasoned flow makes sense to you?

Reply:
Should a fixer changing PHPDoc be risky?

You want to have 2 fixers - as it is in [PHP CS Fixer: custom fixers](https://github.com/kubawerlos/php-cs-fixer-custom-fixers) - if so, then we can discuss the riskiness of the other in separate PR, but first let The One decide - @keradus - single, configurable fixer as in this PR or 2 separate fixers as you can see in [PHP CS Fixer: custom fixers](https://github.com/kubawerlos/php-cs-fixer-custom-fixers)?

Reply:
Yes, even "maybe incorrect" changes on PHPDoc must be risky.

Reply:
Personally I don't consider it as risky because it does not break the code (it modifies _comment_, even if that comment is a part of a contract). The only thing that can be broken is SA in the QA process, but by default this fixer should make 1:1 change (from `x[]` to `array<x>`) and when someone opts-in for `list` the result is also expected. It can be used for detecting non-list arrays and allow fixing them, just oike @kubawerlos did in the other PR.

I don't see the need for 2 fixers here and it was me who opted for a single, configurable fixer in private chat with @kubawerlos.

Reply:
PHPDoc should be considered part of code. SA is part of code. reflection is part of code. So thus if PHPDoc change is risky, the fixer must be risky. Thus I belive two fixers would be better as `xxx[]` to `array<xxx>` is not risky, but `array<xxx>` to `list<xxx>` is risky.

Reply:
@mvorisek every PHPStan update can cause new errors in SA, how's that different from making explicit decision about `list` when using this fixer? You do both on purpose. I wouldn't be too strict when it comes to riskiness of phpDoc changes... But the last word goes to @keradus anyway.

Reply:
my attention was grabbed outside of GH to look at this particular comment. I was not having a chance to look into all threads in this PR.

> Should a fixer changing PHPDoc be risky?
> 
> You want to have 2 fixers - as it is in [PHP CS Fixer: custom fixers](https://github.com/kubawerlos/php-cs-fixer-custom-fixers) - if so, then we can discuss the riskiness of the other in separate PR, but first let The One decide - @keradus - single, configurable fixer as in this PR or 2 separate fixers as you can see in [PHP CS Fixer: custom fixers](https://github.com/kubawerlos/php-cs-fixer-custom-fixers)?

Riskiness is sth that can change the meaning. If we change type (regardless it's declared in code, phpdoc or attribute) to type that may have other meaning, it's risky.
With that, I would, myself, go with 2 fixers:
- one converting array to list-or-iterable (maybe array to list is enough), eg `array<int>` -> `list<int>`, but not `array<string, int> nor `array{string, string, string}`). this is Risky.
- one converting array syntax - and TBH I wonder if we need to support each possible conversion, ie convert `Foo[]` to `array<mixed, Foo>` (not sure if I would go with `array<Foo>` as it is not. Probably yes to allow later risky conversion to `list<Foo>`)